plugins/formal-verification/formver.uniqueness/src/org/jetbrains/kotlin/formver/UniqueCFA.kt

```diff
@@ -0,0 +1,171 @@
+            // TODO: No implemented yet, for now just return from one branch
+            return a
+        }
+
+        override fun visitEdge(
+            from: CFGNode<*>,
+            to: CFGNode<*>,
+            metadata: Edge,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            val dataForEdge = super.visitEdge(from, to, metadata, data)
+            return dataForEdge
+        }
+
+        override fun visitNode(
+            node: CFGNode<*>,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            if (node is EnterNodeMarker) {
+                context.uniqueStack.add(ArrayDeque())
+            } else if (node is ExitNodeMarker) {
+                context.uniqueStack.popLast().lastOrNull()?.let { context.uniqueStack.last().add(it) }
+            }
+            return data
+        }
+
+        override fun visitFunctionEnterNode(
+            node: FunctionEnterNode,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            val dataForNode = visitNode(node, data)
+            val declaration = node.fir
+
+            // construct context and start control flow analysis
+            if (declaration.receiverParameter != null) {
+                TODO("Ignore receiver for now, not sure how to convert it into property symbol yet")
+            }
+            val valueParameters =
+                declaration.valueParameters.associate {
+                    Pair<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>(
+                        it.symbol,
+                        if (it.hasAnnotation(this.context.uniqueId, this.context.session)) {
+                            setOf(UniqueLevel.Unique)
+                        } else {
+                            setOf(UniqueLevel.Shared)
+                        }
+                    )
+                }.toMap()
+            return dataForNode.transformValues { it.putAll(valueParameters) }
+        }
+
+        override fun visitVariableDeclarationNode(
+            node: VariableDeclarationNode,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            val dataForNode = visitNode(node, data)
+            val lSymbol = node.fir.symbol
+            if (node.fir.initializer == null) {
+                return dataForNode.transformValues { it.put(lSymbol, setOf(UniqueLevel.Shared)) }
+            }
+            val rhsUnique = when (val last = context.uniqueStack.last().last()) {
```

Comment:
Stuff like this is very hard to read; please wrap it in some kind of easy-to-work-with interface.